### DISCUSSION TITLE
Subscription not working the same on v4 as it used to be on v3

### DISCUSSION THREAD

**Discussion by JumBay** [urql-graphql/urql#3189] [Q&A] [answered]
Hi,

If you look at this code: 
```ts
import { Client, fetchExchange, gql, subscriptionExchange } from '@urql/core';
import { SubscribePayload, createClient } from 'graphql-ws';
import { pipe, toObservable } from 'wonka';

const wsEndpoint = 'ws://localhost:8080/v1/graphql';
const httpEndpoint = 'http://localhost:8080/v1/graphql';

const accessToken = 'MY_ACCESS_TOKEN';

const headers = {
  headers: {
    'X-Hasura-Role': 'user',
    Authorization: `Bearer ${accessToken}`,
  },
};

const documentNode = gql`
  subscription LatestCustomer {
    affiliation_customer {
      id
    }
  }
`;

const wsClient = createClient({
  url: wsEndpoint,
  lazy: true,
  connectionParams: () => {
    return headers;
  },
});

const client = new Client({
  url: httpEndpoint,
  exchanges: [
    fetchExchange,
    subscriptionExchange({
      forwardSubscription: (request) => ({
        subscribe: (sink) => ({
          unsubscribe: wsClient.subscribe(request as SubscribePayload, sink),
        }),
      }),
    }),
  ],
  fetchOptions: () => headers,
});

const sub = client.subscription(documentNode, {});

const subLastCustomer = pipe(sub, toObservable);

subLastCustomer.subscribe(({ data }) => {
  console.log('LatestCustomerDocument 1');
});

setTimeout(() => {
  console.log('setTimeout called');

  subLastCustomer.subscribe(({ data }) => {
    console.log('LatestCustomerDocument 2');
  });
}, 2000);

```

On urql v3 it was running the request twice, but in v4 the second request is not executed at all. 

How is it possible to do this on v4 as it used to be?

**Comment by kitten** [comment]
The `dedupExchange` logic, or rather, deduplication in general, is now enforced in the `Client` and “built-in” so to speak, so there’s not an exact way of achieving this anymore.

Subscriptions and Queries are generally regarded to be side-effectless however, hence the change to enforce this and get rid of some boilerplate and edge-cases.

However, since you have two active subscriptions with identical variables those are regarded as redundant and identical. Is there any specific use-case you have for this or something specific you’re trying to achieve on an API or client-side here though?

**Reply by JumBay** [reply]
Thanks for your reply.

Actually I'm using this in an angular application with an angular form (for filters) to display some stats in real time with the following code:

```ts
@Component({
  templateUrl: './stats-page.component.html',
})
export class StatsPageComponent {
...

  latestCustomers$ = this.form.controls.platform.valueChanges.pipe(
    startWith(this.form.controls.platform.getRawValue()),
    switchMap((platform) =>
      this.graphQLClient.subscription({
        query: LatestCustomerDocument,
        variables: {
          platform: (platform === 'all' ? {} : { _eq: platform }) as PlatformEnumComparisonExp,
        },
      })
    ),
    map(({ data, error }) => {

      if (error) {
        throw error;
      }

      return data?.affiliation_customer || [];
    })
  );

}
```
When the user enters in that page, the query is called, but if he leaves the page and comes back again, the same query is not called again due to the deduplication I guess. At least I would expect to get the previous/cached value or been able to force the query to rerun.

**Reply by kitten** [reply]
Yep, In case of a query, it would indeed be re-run, however, if you look at subscriptions, they typically don't always represent updating live data, but also events at times (hence the need for an `@live` query spec), so, having the same subscription active twice is often unintended behaviour.

> When the user enters in that page, the query is called, but if he leaves the page and comes back again, the same query is not called again due to the deduplication I guess.

This sounds to me like you potentially have a leaky stream/observable active here 🤔 

What's supposed to happen (which you can validate with the `debugExchange`) is that if the query/subscription stream terminate, a `teardown` event is issued, which then in turn tells the `Client` that there aren't any more subscribers.

However, if this doesn't happen, in particular for your subscription operation, the subscription will just remain active in the background, and once you come back to the page, it'll only pick up new events — if anything.

Without seeing the rest of this, I obviously can't validate all of this, but I'd suspect that something is not terminating properly on navigation 🤔

**Reply by JumBay** [reply]
Ok I've made a simple example and I've managed to make this example works but not as I would expect. 
Here is the code:
```ts
import { Component, OnDestroy, OnInit } from '@angular/core';
import { GraphQLClient } from '@partners/ngx-urql';
import { LatestCustomerDocument } from './operations.auth.generated';

@Component({
  template: '<div></div>',
})
export class StatsPageComponent implements OnInit, OnDestroy {
  subscription!: any;

  constructor(private readonly graphQLClient: GraphQLClient) {}

  ngOnInit() {
    this.subscription = this.graphQLClient
      .subscription({
        query: LatestCustomerDocument,
        variables: {
          platform: {},
        },
      })
      .subscribe(({ data }) => {
        console.log('LatestCustomerDocument', { data });
      });
  }

  ngOnDestroy(): void {
    this.subscription.unsubscribe();
  }
}
```

And here is the `this.graphQLClient.subscription` method:

```ts
import { Observable } from 'rxjs'; 
import { pipe, toObservable } from 'wonka';

  public subscription<Data, Variables extends AnyVariables = AnyVariables>(
    args: OperationArguments<Data, Variables>
  ): Observable<OperationResult<Data, Variables>> {
    return new Observable((observer: any) => {
      pipe(
        this.client.subscription<Data, Variables>(args.query, args.variables as any, args.context),
        toObservable
      ).subscribe(observer);
    });
  }
```
We used to wrap this with the `pipe()` method from wonka to get an observable, but if I only do this it works
```ts
  public subscription<Data, Variables extends AnyVariables = AnyVariables>(args: OperationArguments<Data, Variables>) {
    return this.client.subscription<Data, Variables>(args.query, args.variables as any, args.context);
  }
```
But this is not an observable anymore.
If I do this:
```ts
  public subscription<Data, Variables extends AnyVariables = AnyVariables>(args: OperationArguments<Data, Variables>) {
    return new Observable((observer: any) => {
      this.client.subscription<Data, Variables>(args.query, args.variables as any, args.context).subscribe(observer);
    });
  }
```
I got wonka error:
```txt
re.mjs:8416 ERROR Error: Uncaught (in promise): TypeError: r is not a function
TypeError: r is not a function
    at wonka.mjs:1064:9
    at Array.<anonymous> (wonka.mjs:332:7)
    at wonka.mjs:422:17
    at wonka.mjs:295:9
    at wonka.mjs:318:9
    at wonka.mjs:742:11
 ```

I guess when we were converting it to an observable via wonka, the unsubscribe doesn't unsubscribe the query, but only the observable

**Reply by kitten** [reply]
If you want to convert a source to an observable you're likely better off using `toObservable` from `wonka`, yea. `subscribe(observer)` doesn't work because `observer` is an Observer object not an `onNext`/`onResult` function

**Reply by JumBay** [reply]
Yeah that's what I'm trying to do, but if I replace this:
```ts
    this.subscription = this.graphQLClient
      .subscription({
        query: LatestCustomerDocument,
        variables: {
          platform: {},
        },
      })
      .subscribe(({ data }) => {
        console.log('LatestCustomerDocument', { data });
      });
```
By 
```ts
 this.subscription = new Observable((observer) => {
      pipe(
        this.graphQLClient.subscription({
          query: LatestCustomerDocument,
          variables: {
            platform: {},
          },
        }),
        toObservable
      ).subscribe(observer);
    }).subscribe((data) => {
      console.log('LatestCustomerDocument', { data });
    });
```
It seems that `this.subscription.unsubscribe();` is only unsubscribing the rxjs/wonka observable but not the source. 
Do you know how to change this?

Thanks

**Reply by JumBay** [reply, marked as answer]
Ok I've managed to make this work without using wonka like this:
```ts
this.subscription = new Observable((observer) => {
      const sub = this.graphQLClient
        .subscription({
          query: LatestCustomerDocument,
          variables: {
            platform: {},
          },
        })
        .subscribe((data) => observer.next(data));

      return {
        unsubscribe: sub.unsubscribe,
      };
    }).subscribe((data) => {
      console.log('LatestCustomerDocument', { data });
    });
```

Thanks again for your help
